examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/CastingApp-JNI.cpp

```diff
@@ -0,0 +1,132 @@
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+
+#include "CastingApp-JNI.h"
+
+#include "../JNIDACProvider.h"
+#include "../support/ErrorConverter-JNI.h"
+#include "../support/RotatingDeviceIdUniqueIdProvider-JNI.h"
+#include "core/CastingApp.h" // from tv-casting-common
+
+#include <app/clusters/bindings/BindingManager.h>
+#include <app/server/Server.h>
+#include <jni.h>
+#include <lib/support/JniReferences.h>
+#include <lib/support/JniTypeWrappers.h>
+
+using namespace chip;
+
+#define JNI_METHOD(RETURN, METHOD_NAME) extern "C" JNIEXPORT RETURN JNICALL Java_com_matter_casting_core_CastingApp_##METHOD_NAME
+
+namespace matter {
+namespace casting {
+namespace core {
+
+CastingAppJNI CastingAppJNI::sInstance;
+
+jobject extractJAppParameter(jobject jAppParameters, const char * methodName, const char * methodSig);
+
+JNI_METHOD(jobject, finishInitialization)(JNIEnv *, jobject, jobject jAppParameters)
+{
+    chip::DeviceLayer::StackLock lock;
+    ChipLogProgress(AppServer, "JNI_METHOD CastingAppJNI.finishInitialization called");
+    VerifyOrReturnValue(jAppParameters != nullptr, support::createJMatterError(CHIP_ERROR_INVALID_ARGUMENT));
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    jobject jUniqueIdProvider =
+        extractJAppParameter(jAppParameters, "getRotatingDeviceIdUniqueIdProvider", "()Lcom/matter/casting/support/DataProvider;");
+    VerifyOrReturnValue(jUniqueIdProvider != nullptr, support::createJMatterError(CHIP_ERROR_INCORRECT_STATE));
+    support::RotatingDeviceIdUniqueIdProviderJNI * uniqueIdProvider = new support::RotatingDeviceIdUniqueIdProviderJNI();
+    err                                                             = uniqueIdProvider->Initialize(jUniqueIdProvider);
+    VerifyOrReturnValue(err == CHIP_NO_ERROR, support::createJMatterError(CHIP_ERROR_INVALID_ARGUMENT));
+
+    // set the RotatingDeviceIdUniqueId
+#if CHIP_ENABLE_ROTATING_DEVICE_ID
+    chip::MutableByteSpan * uniqueId = uniqueIdProvider->Get();
+    if (uniqueId != nullptr)
+    {
+        chip::DeviceLayer::ConfigurationMgr().SetRotatingDeviceIdUniqueId(*uniqueId);
+    }
+#endif // CHIP_ENABLE_ROTATING_DEVICE_ID
+
+    // get the DACProvider
+    jobject jDACProvider = extractJAppParameter(jAppParameters, "getDacProvider", "()Lcom/matter/casting/support/DACProvider;");
+    VerifyOrReturnValue(jDACProvider != nullptr, support::createJMatterError(CHIP_ERROR_INCORRECT_STATE));
+
+    // set the DACProvider
+    JNIDACProvider * dacProvider = new JNIDACProvider(jDACProvider);
+    chip::Credentials::SetDeviceAttestationCredentialsProvider(dacProvider);
+
+    return support::createJMatterError(CHIP_NO_ERROR);
+}
+
+JNI_METHOD(jobject, finishStartup)(JNIEnv *, jobject)
+{
+    chip::DeviceLayer::StackLock lock;
+    ChipLogProgress(AppServer, "JNI_METHOD CastingAppJNI.finishStartup called");
+    auto & server = chip::Server::GetInstance();
+
+    // TODO: Set AppDelegate
```

Comment:
Are these TODO's optional? Just wondering if this means there is missing functionality right now.

Reply:
Not optional. I'll be refactoring some of more the old code into new classes in following PRs, and will then complete these TODOs.